src/rgw/rgw_notify.cc

```diff
@@ -179,25 +181,88 @@ class Manager {
     }
   }
 
+  // clear stale reservations on a queue if next cleanup time reached
+  // return the number of clearted reservations and update the next cleanup time
+  unsigned clean_stale_reservation(std::chrono::time_point<ceph::time_detail::coarse_real_clock>& next_cleanup, 
+      spawn::yield_context yield,
+      const std::string& queue_name) 
+  {
+    const auto now = ceph::coarse_real_time::clock::now();
+    if (now < next_cleanup) {
+      return 0;
+    }
+    const auto stale_timestamp = now - std::chrono::seconds(stale_reservations_period_s);
+    next_cleanup = now + std::chrono::seconds(reservations_cleanup_period_s);
+    const auto tp = ceph::coarse_real_time::clock::to_time_t(next_cleanup);
+    ldout(cct, 20) << "INFO: stale reservation cleanup for queue: " << queue_name 
+      << ". next cleanup will happen at: " << std::ctime(&tp) << dendl;
+    librados::ObjectReadOperation list_op;
+    bufferlist obl;
+    int rval;
+    cls_2pc_queue_list_reservations(list_op, &obl, &rval);
+    auto ret = rgw_rados_operate(rados_ioctx, queue_name, &list_op, nullptr, optional_yield(io_context, yield));
+    if (ret < 0) {
+      ldout(cct, 5) << "WARNING: failed to cleanup stale reservation from queue: " << queue_name
+        << ". failed to get list, error: " << ret << dendl;
+      return 0;
+    }
+
+    cls_2pc_reservations reservations;
+    ret = cls_2pc_queue_list_reservations_result(obl, reservations);
+    if (ret < 0) {
+      ldout(cct, 5) << "WARNING: failed to cleanup stale reservation from queue: " << queue_name
+      << ". failed to parse list, error: " << ret << dendl;
+      return 0;
+    }
+    
+    auto cleaned_reservations = 0U;
+    for (const auto& r : reservations) {
+      if (r.second.timestamp < stale_timestamp) {
+        librados::ObjectWriteOperation abort_op;
+        cls_2pc_queue_abort(abort_op, r.first);
```

Comment:
it would be nice to be able to do these in batches. better yet, maybe we could even avoid `cls_2pc_queue_list_reservations()` in the first place by providing a new op like `cls_2pc_queue_expire_reservations()` that took this `stale_timestamp` as an argument. you could read the reservations, remove everything that expired, and write it back in a single cls call

Reply:
not sure if batching would work.
had issues with batching before, since the abort action both read and write from the queue.
adding it as a construct to cls makes sense.